kong/plugins/bot-detection/handler.lua

```diff
@@ -19,10 +44,79 @@ local function get_user_agent()
   return user_agent
 end
 
+local function examine_agent(user_agent, conf)
+  user_agent = strip(user_agent)
+
+  if conf.whitelist then
+    for _, rule in ipairs(conf.whitelist) do
+      if re_match(user_agent, rule) then
+        return MATCH_WHITELIST
+      end
+    end
+  end
+
+  if conf.blacklist then
+    for _, rule in ipairs(conf.blacklist) do
+      if re_match(user_agent, rule) then
+        return MATCH_BLACKLIST
+      end
+    end
+  end
+
+  for _, rule in ipairs(rules.bots) do
+    if re_match(user_agent, rule) then
+      return MATCH_BOT
+    end
+  end
+
+  return MATCH_EMPTY
+end
+
 function BotDetectionHandler:new()
   BotDetectionHandler.super.new(self, "bot-detection")
 end
 
+function BotDetectionHandler:init_worker()
+  local singletons    = require "kong.singletons"
+  local worker_events = singletons.worker_events
+  local dao_factory   = singletons.dao
+
+  -- load our existing plugins to create our cache spaces
+  local plugins, err = dao_factory.plugins:find_all({
+    name = "bot-detection",
+  })
+  if err then
+    ngx.log(ngx.ERR, "err in fetching plugins: ", err)
+  end
+
+  for i = 1, #plugins do
+    ua_cache[plugins[i].api_id] = {}
+  end
+
+  -- catch updates and tell each worker to adjust their cache
+  -- accordingly. for create/updates we create/purge the cache
+  -- by simply creating a new lua table; for deletions, as we no
+  -- longer need this cache data, we just remove it
+  worker_events.register(function(data)
+    if data.entity.name == "bot-detection" then
+      worker_events.post("bot-detection-invalidate", data.operation,
```

Comment:
My bad, the original CRUD event is actually a local event (https://github.com/Mashape/kong/blob/next/kong/core/handler.lua#L126). I overlooked this but now remember the reason: for plugins subscribing to the event to avoid duplicating the event handling across all workers if undesired (propagation to other workers is deferred to the plugin itself as this implementation actually demonstrates).